src/irc/core/modes.c

```diff
@@ -498,7 +500,82 @@ static void event_mode(IRC_SERVER_REC *server, const char *data,
 			parse_channel_modes(chanrec, nick, mode, TRUE);
 	}
 
-	g_free(params);
+	g_free(channel);
+	g_free(mode);
+}
+
+// Given <params> to a MODE command, parse them into the target of the modes (a
+// channel or nickname) and a string holding the modes and their parameters.
+//
+// ":" will be removed from any <trailing> parameter.
+//
+// Since the parsing combines several parameters into one string, the exact
+// position of <trailing> is not particularly important. This is not very RFC
+// correct, but it has historically worked.
+//
+// For example:
+//
+// "#channel +o test" gets parsed to: target = "#channel" and modes = "+o test"
+// "#channel :+o test" gets parsed to: target = "#channel" and modes = "+o test"
+// "#channel +o :test" gets parsed to: target = "#channel" and modes = "+o test"
+//
+// The caller must free the target and modes strings.
+void get_mode_params(char const * const params, char * * target, char * * modes)
+{
+	char * params_copy = NULL;
+	char * params_ptr = NULL;
+	char * s = NULL;
+	char * mode_buf = NULL;
+
+	g_return_if_fail(params != NULL);
+	g_return_if_fail(target != NULL);
+	g_return_if_fail(modes != NULL);
+
+	// Copy params as parsing modifies the buffer.
+	params_copy = g_strdup(params);
+	params_ptr = params_copy;
+
+	// Get the target.
+
+	while (*params_ptr == ' ') {
+		params_ptr++;
+	}
+
+	s = event_get_param(&params_ptr);
+	if (s == NULL || strlen(s) == 0) {
+		g_free(params_copy);
+		return;
+	}
+	*target = g_strdup(s);
+
+	// Get the mode string. This is all remaining parameters as a single string.
+
+	// This is potentially larger than we need.
+	mode_buf = g_malloc0(strlen(params_ptr)+1);
+
+	// Get a single parameter at a time so that we correctly handle ":".
+	while (strlen(params_ptr) > 0) {
+		// event_get_param() doesn't ignore leading whitespace. Since parameters
+		// can be separated by more than a single space, strip spaces before using
+		// it.
+		while (*params_ptr == ' ') {
+			params_ptr++;
+		}
+
+		s = event_get_param(&params_ptr);
+		if (s == NULL || strlen(s) == 0) {
+			break;
+		}
+
+		if (strlen(mode_buf) > 0) {
+			strcat(mode_buf, " ");
+		}
+		strcat(mode_buf, s);
```

Comment:
Use GString instead of g_malloc0 / strcat / etc

https://developer.gnome.org/glib/stable/glib-Strings.html